internal/jsonwall/jsonwall.go

```diff
@@ -0,0 +1,307 @@
+//                     }
+//             }
+//     }
+//
+package jsonwall
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"sort"
+	"strings"
+)
+
+// DBWriter holds in memory the state of a database while it's being prepared
+// for serialization and implements the WriterTo interface for assembling it.
+type DBWriter struct {
+	options *DBWriterOptions
+	entries [][]byte
+}
+
+type DBWriterOptions struct {
+	// Schema is included in the database header to help the decoding
+	// process. The value is made available when reading, and is not
+	// internally interpreted.
+	Schema string
+}
+
+// NewDBWriter returns a database writer that can assemble new databases.
+func NewDBWriter(options *DBWriterOptions) *DBWriter {
+	if options == nil {
+		options = &DBWriterOptions{}
+	}
+	return &DBWriter{options: options}
+}
+
+// Add encodes the provided value as a JSON object and includes the resulting
+// data into the database being created.
+func (dbw *DBWriter) Add(value any) error {
+	data, err := json.Marshal(value)
+	if err != nil {
+		return err
+	}
+	if len(data) == 0 || data[0] != '{' {
+		return fmt.Errorf("invalid database value: %#v", value)
+	}
+	dbw.entries = append(dbw.entries, data)
+	return nil
+}
+
+type sortableEntries [][]byte
+
+func (e sortableEntries) Less(i, j int) bool { return bytes.Compare(e[i], e[j]) < 0 }
+func (e sortableEntries) Swap(i, j int)      { e[i], e[j] = e[j], e[i] }
+func (e sortableEntries) Len() int           { return len(e) }
+
+type jsonwallHeader struct {
+	Version string `json:"jsonwall"`
+	Schema  string `json:"schema,omitempty"`
+	Count   int    `json:"count"`
+}
+
+const jsonwallVersion = "1.0"
+
+func (dbw *DBWriter) writeHeader(w io.Writer, count int) (int, error) {
+	data, err := json.Marshal(&jsonwallHeader{
+		Version: jsonwallVersion,
+		Schema:  dbw.options.Schema,
+		Count:   count,
+	})
+	if err != nil {
+		return 0, fmt.Errorf("internal error: cannot marshal database header")
+	}
+	return w.Write(append(data, '\n'))
+}
+
+// WriteTo assembles the current database state and writes it to w.
+func (dbw *DBWriter) WriteTo(w io.Writer) (n int64, err error) {
+	m, err := dbw.writeHeader(w, len(dbw.entries)+1)
+	n += int64(m)
+	sort.Sort(sortableEntries(dbw.entries))
+	if err != nil {
+		return n, err
+	}
+	for _, entry := range dbw.entries {
+		m, err := w.Write(entry)
+		n += int64(m)
+		if err == nil {
+			m, err = w.Write([]byte{'\n'})
+			n += int64(m)
+		}
+		if err != nil {
+			return n, err
+		}
+	}
+	return n, nil
+}
+
+// ReadDB reads into memory the database from the provided r.
+func ReadDB(r io.Reader) (*DB, error) {
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+	record := 0
+	for i := range data {
+		if data[i] == '\n' {
+			record = i + 1
+			break
+		}
+	}
+	var header jsonwallHeader
+	err = json.Unmarshal(data[:record], &header)
+	if err != nil {
+		return nil, fmt.Errorf("invalid database content")
```

Comment:
Same as Fred commented below. Can the error be more specific?
```
return nil, fmt.Errorf("invalid database header: %w")
```

Reply:
That was intentional because if even the header bytes are wrong, that file is completely corrupted, or perhaps more likely, it's not a jsonwall file at all. Including a bogus attempt to parse binary data as JSON, for example, is not helpful. This is similar to reading the first 4 bytes of magic from some file format and saying "nope, not a JPG" for instance.

Reply:
I understand what you mean, but I thought that the format is supposed to be transparent about its JSON-ness and users will be able (and encouraged) to inspect it with `jq` (and perhaps `zstdcat`). Of course, this doesn't mean we will support the files written by anything other than chisel.

Reply:
That's true, but what I said doesn't seem to disagree with that. Again, think about the case of the JPG image: when the application says "that JPG image looks bogus" instead of "that image is 10 billion pixels wide", it's not hiding its "JPEGness", it's simply trying to present a more pleasant error message when the file looks completely wrong.